activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -89,34 +89,51 @@ def self.quoted_table_names # :nodoc:
         @quoted_table_names ||= {}
       end
 
-      def initialize(connection, logger = nil, config = {}) # :nodoc:
+      def initialize(config_or_deprecated_connection, deprecated_logger = nil, deprecated_connection_options = nil, deprecated_config = nil) # :nodoc:
         super()
 
-        @raw_connection      = connection
-        @owner               = nil
-        @instrumenter        = ActiveSupport::Notifications.instrumenter
-        @logger              = logger
-        @config              = config
-        @pool                = ActiveRecord::ConnectionAdapters::NullPool.new
-        @idle_since          = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+        if config_or_deprecated_connection.is_a?(Hash)
+          @raw_connection = nil
+          @config = config_or_deprecated_connection.symbolize_keys
+          @logger = ActiveRecord::Base.logger
+
+          if deprecated_logger || deprecated_connection_options || deprecated_config
+            raise ArgumentError, "new API accepts just one config hash"
```

Comment:
This is a confusing error because if I passed a config hash and a logger I wouldn't know what this error was indicating I should change. I know this is a WIP but wanted to note I think each deprecated option should have it's own error to be clearer about the new API.